src/vcf/assembly_checker.cpp

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+
+#include "vcf/assembly_checker.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+    namespace assembly_checker
+    {
+
+      bool check_vcf_ref(std::istream &vcf_input,
+                         std::istream &fasta_input,
+                         std::istream &fasta_index_input,
+                         std::vector<std::unique_ptr<ebi::vcf::AssemblyReportWriter>> &outputs)
+      {
+          std::vector<char> vector_line;
+          vector_line.reserve(default_line_buffer_size);
+          std::set<std::string> absent_chromosomes;
+
+          // Reading FASTA index, and querying FASTA file
+          auto index = bioio::read_fasta_index(fasta_index_input);
+
+          while (util::readline(vcf_input, vector_line).size() != 0) {
+              std::string line{vector_line.begin(), vector_line.end()};
+              std::vector<std::string> record_columns;
+
+              if (boost::starts_with(line, "#")) {
+                  continue;
+              }
+
+              VcfVariant vcf_variant{line};
+
+              if (index.count(vcf_variant.chromosome) == 0) {
+                  absent_chromosomes.insert(vcf_variant.chromosome);
+                  continue;
+              }
+
+              if (vcf_variant.position == 0) {
+                  BOOST_LOG_TRIVIAL(warning) << "Position 0 should only be used for a telomere";
+                  continue;
+              }
+
+              auto fasta_sequence = bioio::read_fasta_contig(fasta_input,
+                                                             index.at(vcf_variant.chromosome),
+                                                             vcf_variant.position - 1,
+                                                             vcf_variant.reference_allele.length());
+              auto reference_sequence = vcf_variant.reference_allele;
+
+              bool match_result = is_matching_sequence(fasta_sequence, reference_sequence);
+
+              for (auto &output : outputs ) {
+                  if (match_result) {
+                      output->write_match(vcf_variant);
+                  } else {
+                      output->write_mismatch(vcf_variant);
+                  }
+              }
+          }
+
+          outputs[0]->finish_report();
+          check_missing_chromosomes(absent_chromosomes);
+          return outputs[0]->is_valid_report();
+      }
+
+      bool is_matching_sequence(std::string fasta_sequence, std::string reference_sequence)
+      {
+          std::transform(fasta_sequence.begin(), fasta_sequence.end(), fasta_sequence.begin(), ::tolower);
+          std::transform(reference_sequence.begin(), reference_sequence.end(), reference_sequence.begin(), ::tolower);
+
+          return fasta_sequence == reference_sequence;
+      }
+
+      void check_missing_chromosomes(std::set<std::string> absent_chromosomes)
+      {
+          if (absent_chromosomes.size() > 0) {
```

Comment:
This will make the process abort if checking against FASTA files split by chromosome. Instead of checking whether any chromosomes are absent, we should check if **all** are.

As long as one is shared across the VCF and FASTA files, then we can run some validation. We can raise a warning is some are not, because it is true that could help to detect mistakes when running the tool; but it should be a logging message instead of an exception.